src/mrpro/operators/WaveletOp.py

```diff
@@ -0,0 +1,374 @@
+        wavelet_name
+            Name of wavelets
+        level
+            Highest wavelet level. If set to None, the highest possible level is calculated based on the domain shape.
+
+        Raises
+        ------
+        ValueError
+            If wavelets are calculated for more than three dimensions.
+        ValueError
+            If wavelet dimensions and domain shape do not match.
+        NotImplementedError
+            If any dimension of the domain shape is odd. Adjoint will lead to the wrong domain shape.
+        """
+        super().__init__()
+        self._domain_shape = domain_shape
+        self._wavelet_name = wavelet_name
+        self._dim = dim
+        self._level = level
+
+        # number of wavelet directions
+        if len(dim) == 1:
+            self.n_wavelet_directions = 1
+        elif len(dim) == 2:
+            self.n_wavelet_directions = 3
+        elif len(dim) == 3:
+            self.n_wavelet_directions = 7
+        else:
+            raise ValueError('Only 1D, 2D and 3D wavelet transforms are supported.')
+
+        if domain_shape is not None:
+            if len(dim) != len(domain_shape):
+                raise ValueError(
+                    'Number of dimensions along which the wavelet transform should be calculated needs to',
+                    'be the same as the domain shape',
+                )
+
+            if any(d % 2 for d in domain_shape):
+                raise NotImplementedError(
+                    'ptwt only supports wavelet transforms for tensors with even number of '
+                    'entries for all considered dimensions.'
+                )
+
+            # size of wavelets
+            wavelet_length = torch.as_tensor((Wavelet(wavelet_name).dec_len,) * len(domain_shape))
+
+            # calculate shape of wavelet coefficients at each level
+            current_shape = torch.as_tensor(domain_shape)
+
+            # if level is None, select the highest possible level.
+            # raise error/warnings if level is not possible or lead to boundary effects
+            verified_level = _check_level(domain_shape, wavelet_length, level)
+
+            if verified_level == 0:  # only a/aa/aaa component possible
+                self.coefficients_shape = [domain_shape]
+            else:
+                self.coefficients_shape = []
+                for _ in range(verified_level):
+                    # Add padding
+                    current_shape = (current_shape / 2).ceil() + wavelet_length // 2 - 1
+                    self.coefficients_shape.extend(
+                        [tuple(current_shape.to(dtype=torch.int64))] * self.n_wavelet_directions
+                    )
+
+                self.coefficients_shape = self.coefficients_shape[::-1]
+                self.coefficients_shape.insert(0, self.coefficients_shape[0])  # shape of a/aa/aaa term
+
+    def forward(self, x: torch.Tensor) -> tuple[torch.Tensor,]:
```

Comment:
To avoid potential bugs and improve code robustness, add explicit type checks for the inputs to the `WaveletOp` class methods, ensuring they meet expected criteria before proceeding with computations. [important]